src/jit/MemoryInl.h

```diff
@@ -1011,3 +1011,557 @@ static void emitStore(sljit_compiler* compiler, Instruction* instr)
+        break;
+    }
+    case ByteCode::I64AtomicRmwOrOpcode: {
+        faddr = GET_FUNC_ADDR(sljit_sw, atomicRmwOr64);
+        break;
+    }
+    case ByteCode::I64AtomicRmwXorOpcode: {
+        faddr = GET_FUNC_ADDR(sljit_sw, atomicRmwXor64);
+        break;
+    }
+    case ByteCode::I64AtomicRmwXchgOpcode: {
+        faddr = GET_FUNC_ADDR(sljit_sw, atomicRmwXchg64);
+        break;
+    }
+    default: {
+        RELEASE_ASSERT_NOT_REACHED();
+        break;
+    }
+    }
+
+    if (addr.memArg.arg != SLJIT_R0) {
+        sljit_emit_op1(compiler, SLJIT_MOV, SLJIT_R0, 0, SLJIT_EXTRACT_REG(addr.memArg.arg), 0);
+    }
+
+    if (srcArgPair.arg1 != SLJIT_MEM1(kFrameReg)) {
+        sljit_emit_op1(compiler, SLJIT_MOV, SLJIT_MEM1(kContextReg), OffsetOfContextField(tmp1) + WORD_LOW_OFFSET, srcArgPair.arg1, srcArgPair.arg1w);
+        sljit_emit_op1(compiler, SLJIT_MOV, SLJIT_MEM1(kContextReg), OffsetOfContextField(tmp1) + WORD_HIGH_OFFSET, srcArgPair.arg2, srcArgPair.arg2w);
+
+        sljit_emit_op2(compiler, SLJIT_ADD, SLJIT_R1, 0, kContextReg, 0, SLJIT_IMM, OffsetOfContextField(tmp1));
+    } else {
+        sljit_emit_op2(compiler, SLJIT_ADD, SLJIT_R1, 0, kFrameReg, 0, SLJIT_IMM, srcArgPair.arg1w - WORD_LOW_OFFSET);
+    }
+
+    if (dstArgPair.arg1 != SLJIT_MEM1(kFrameReg)) {
+        sljit_emit_op2(compiler, SLJIT_ADD, SLJIT_R2, 0, kContextReg, 0, SLJIT_IMM, OffsetOfContextField(tmp2));
+    } else {
+        sljit_emit_op2(compiler, SLJIT_ADD, SLJIT_R2, 0, kFrameReg, 0, SLJIT_IMM, dstArgPair.arg1w - WORD_LOW_OFFSET);
+    }
+
+    sljit_emit_icall(compiler, SLJIT_CALL, SLJIT_ARGS3V(P, P, P), SLJIT_IMM, faddr);
+
+    if (dstArgPair.arg1 != SLJIT_MEM1(kFrameReg)) {
+        sljit_emit_op1(compiler, SLJIT_MOV, dstArgPair.arg1, dstArgPair.arg1w, SLJIT_MEM1(kContextReg), OffsetOfContextField(tmp2) + WORD_LOW_OFFSET);
+        sljit_emit_op1(compiler, SLJIT_MOV, dstArgPair.arg2, dstArgPair.arg2w, SLJIT_MEM1(kContextReg), OffsetOfContextField(tmp2) + WORD_HIGH_OFFSET);
+    }
+}
+
+static void emitAtomicRmwCmpxchg64(sljit_compiler* compiler, Instruction* instr)
+{
+    uint32_t options = MemAddress::CheckNaturalAlignment | MemAddress::AbsoluteAddress;
+    uint32_t size = 8;
+    sljit_u32 offset;
+
+    Operand* operands = instr->operands();
+    MemAddress addr(options, instr->requiredReg(0), instr->requiredReg(1), instr->requiredReg(2));
+    AtomicRmwCmpxchg* rmwCmpxchgOperation = reinterpret_cast<AtomicRmwCmpxchg*>(instr->byteCode());
+    offset = rmwCmpxchgOperation->offset();
+
+    addr.check(compiler, operands, offset, size);
+
+    if (addr.memArg.arg == 0) {
+        return;
+    }
+
+    JITArgPair srcExpectedArgPair(operands + 1);
+    JITArgPair srcValueArgPair(operands + 2);
+    JITArgPair dstArgPair(operands + 3);
+    sljit_s32 type = SLJIT_ARGS3V(P, P, P);
+    sljit_s32 faddr GET_FUNC_ADDR(sljit_sw, atomicRmwCmpxchg64);
+
+    sljit_emit_op1(compiler, SLJIT_MOV, SLJIT_R0, 0, SLJIT_EXTRACT_REG(addr.memArg.arg), 0);
```

Comment:
The problem here is that if R0 is overwritten and used by srcExpectedArgPair or srcValueArgPair, then an incorrect value is written. I would copy the arguments to memory, if needed, and then initialize R0 to R2 after that.